### ISSUE TITLE
Multiple row-of-values bound with parameterised query do not work as expected

### ISSUE THREAD

**Issue by CMa72** [ClickHouse/clickhouse-odbc#324]
Hi,
We are testing ClickHouse using ODBC-driver on Windows 10 (ClickHouse server version 20.10.3.30 on a docker).
We need array-binding (arrayDML) functionality (multiple row-of-values bound with a parameterised query), but it seems, that it does not work as expected.

A simple test with 3 rows of data bound to the INSERT statement ends up with only the first row of data being inserted (expected are three inserted rows).

See the below simple code, that is able to replicate the issue.

First, lets create a table:
```
CREATE TABLE test.drv04_array_dml(
  ID int not null,
  c_event_date date,
  dec28_8 DECIMAL( 28, 8 ),
  dec38_8 DECIMAL( 38, 3 )
) engine = MergeTree PARTITION BY toYYYYMM(c_event_date ) ORDER BY (ID, c_event_date)  SETTINGS index_granularity = 8192;
```

And here is the code:

```
#include <stdio.h>
#if defined(_MSC_VER)
#  include <windows.h>
#  include <tchar.h>
#endif
#include <sql.h>
#include <sqlext.h>

SQLHENV     g_hEnvironment;
SQLHDBC     g_hConnection;
SQLHSTMT    g_hStmt;

void print_error(
   SQLSMALLINT       a_nHandleType,
   SQLHANDLE         a_hHandle,
   SQLRETURN         a_nReturned,
   const char*       ac_szWhere
) {
   SQLINTEGER nNtvErrCode;
   SQLCHAR txtNtvErrMsg[SQL_MAX_MESSAGE_LENGTH];
   SQLSTATE txtSQLState;
   SQLSMALLINT nNtvErrMsg_len;
   SQLSMALLINT nMessageNo = 1;
   int nThereAreMore = 1;

   printf( "\n!!\n" );
   printf( "!! error occurred, details ... \n" );
   printf( "!!   error code == %d, where == %s\n", a_nReturned, ac_szWhere );
   printf( "!!\n" );

   do {
      switch( SQLGetDiagRec( a_nHandleType, a_hHandle, nMessageNo++, txtSQLState, &nNtvErrCode, txtNtvErrMsg, SQL_MAX_MESSAGE_LENGTH, &nNtvErrMsg_len ) ) {
         case SQL_SUCCESS: {
            // txtSQLState is five-character SQLSTATE code (according to docs, it should be terminated with NULL, but on linux it is not)
            txtSQLState[SQL_SQLSTATE_SIZE] = '\0';

            printf( "!! nMessageNo : %d ---\n", nMessageNo - 1 );
            printf( "!! SQLSTATE   : %s\n", txtSQLState );
            printf( "!! ntvErrCode : %ld\n", ( long ) nNtvErrCode );
            printf( "!! NtvErrorMsg: %s\n", txtNtvErrMsg );
            break;
         }
         default:
         case SQL_NO_DATA: {
            nThereAreMore = 0;
            break;
         }
      }
   } while( nThereAreMore );
}

void print_error_and_exit(
   SQLSMALLINT       a_nHandleType,
   SQLHANDLE         a_hHandle,
   SQLRETURN         a_nReturned,
   const char* ac_szWhere
) {
   print_error( a_nHandleType, a_hHandle, a_nReturned, ac_szWhere );
   exit( -1 );
}

void print_error_and_exit_conn( SQLRETURN a_nReturned, const char* ac_szWhere )
{
   print_error_and_exit( SQL_HANDLE_DBC, g_hConnection, a_nReturned, ac_szWhere );
}

void print_error_and_exit_env( SQLRETURN a_nReturned, const char* ac_szWhere )
{
   print_error_and_exit( SQL_HANDLE_ENV, g_hEnvironment, a_nReturned, ac_szWhere );
}

void print_error_and_exit_stmt( SQLHSTMT a_hStmt, SQLRETURN a_nReturned, const char* ac_szWhere )
{
   print_error_and_exit( SQL_HANDLE_STMT, a_hStmt, a_nReturned, ac_szWhere );
}

int main( int argc, char** argv ) {

   SQLCHAR szDSN[] = "test";
   SQLCHAR szUser[] = "test";
   SQLCHAR szPassword[] = "test";
   SQLCHAR szStmt[] = "INSERT INTO test.drv04_array_dml (ID, dec38_8, dec28_8 ) VALUES ( ?, ?, ? )";

   SQLRETURN nReturned = SQLAllocHandle( SQL_HANDLE_ENV, SQL_NULL_HANDLE, &g_hEnvironment );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_env( nReturned, "handle alloc" );

   nReturned = SQLSetEnvAttr( g_hEnvironment, SQL_ATTR_ODBC_VERSION, ( SQLPOINTER ) SQL_OV_ODBC3, 0 );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_env( nReturned, "env handle alloc" );

   nReturned = SQLAllocHandle( SQL_HANDLE_DBC, g_hEnvironment, &g_hConnection );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_env( nReturned, "dbc handle alloc" );

   nReturned = SQLConnect( g_hConnection, szDSN, SQL_NTS, szUser, SQL_NTS, szPassword, SQL_NTS );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_conn( nReturned, "connection establishing" );

   int nP1_buffer[3] = { 1, 2, 3 };
   SQLLEN nP1_arrIndOrLen[3] = { 0, 0, 0 };
   SQLCHAR chP2_buffer[3][10] = { "1", "2", "3" };
   SQLLEN nP2_arrIndOrLen[3] = { SQL_NTS, SQL_NTS, SQL_NTS };
   SQLCHAR chP3_buffer[3][10] = { "1", "2", "3" };
   SQLLEN nP3_arrIndOrLen[3] = { SQL_NTS, SQL_NTS, SQL_NTS };

   nReturned = SQLAllocHandle( SQL_HANDLE_STMT, g_hConnection, &g_hStmt );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_conn( nReturned, "SQLAllocHandle SQL_HANDLE_STMT" );

   nReturned = SQLPrepare( g_hStmt, ( SQLCHAR* ) ( szStmt ), SQL_NTS );
   if( nReturned != SQL_SUCCESS && nReturned != SQL_SUCCESS_WITH_INFO )
      print_error_and_exit_stmt( g_hStmt, nReturned, "SQLPrepare" );

   nReturned = SQLSetStmtAttr( g_hStmt, SQL_ATTR_PARAMSET_SIZE, ( SQLPOINTER ) 3, 0 );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( g_hStmt, nReturned, "SQLSetStmtAttr" );

   nReturned = SQLBindParameter(
      g_hStmt, 1, SQL_PARAM_INPUT, SQL_INTEGER, SQL_INTEGER, 10, 10,
      nP1_buffer, -1, nP1_arrIndOrLen
   );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( g_hStmt, nReturned, "SQLBindParameter p1" );

   nReturned = SQLBindParameter(
      g_hStmt, 2, SQL_PARAM_INPUT, SQL_C_CHAR, SQL_DECIMAL, 38, 8,
      chP2_buffer, 10, nP2_arrIndOrLen
   );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( g_hStmt, nReturned, "SQLBindParameter p2" );

   nReturned = SQLBindParameter(
      g_hStmt, 3, SQL_PARAM_INPUT, SQL_C_CHAR, SQL_DECIMAL, 28, 8,
      chP3_buffer, 10, nP3_arrIndOrLen
   );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( g_hStmt, nReturned, "SQLBindParameter p3" );

   nReturned = SQLExecute( g_hStmt );
   if( nReturned != SQL_SUCCESS ) {
      print_error( SQL_HANDLE_STMT, g_hStmt, nReturned, "SQLExecute" );
      nReturned = SQLFreeHandle( SQL_HANDLE_STMT, g_hStmt );
      return 1;
   }

   // free the handles (code omitted)

   return 0;
}
```

The above code always inserts only the 1st row:

```
:) select * from test.drv04_array_dml;
┌─ID─┬─c_event_date─┬────dec28_8─┬─dec38_8─┐
│  1 │   1970-01-01 │ 1.00000000 │   1.000 │
└────┴──────────────┴────────────┴─────────┘
```

while expected are three rows, like below:

```
:) select * from test.drv04_array_dml;
┌─ID─┬─c_event_date─┬────dec28_8─┬─dec38_8─┐
│  1 │   1970-01-01 │ 1.00000000 │   1.000 │
└────┴──────────────┴────────────┴─────────┘
┌─ID─┬─c_event_date─┬────dec28_8─┬─dec38_8─┐
│  2 │   1970-01-01 │ 2.00000000 │   2.000 │
└────┴──────────────┴────────────┴─────────┘
┌─ID─┬─c_event_date─┬────dec28_8─┬─dec38_8─┐
│  3 │   1970-01-01 │ 3.00000000 │   3.000 │
└────┴──────────────┴────────────┴─────────┘

3 rows in set. Elapsed: 16.9981ms
```

Maybe am doing something wrong, but it has been successfully tested on other DBMS and its odbc-driver ... could you check this please ?

Regards,
Marek Ćwikliński

**Comment by traceon** [comment]
I believe, after `SQLExecute` you need to call one `SQLMoreResults(...)` for each next parameter value set (and fetch data in between those calls for each respective result set).

**Comment by CMa72** [comment]
Hi,

Thank you for your suggestion.
I used one SQLMoreResults(...) for each next parameter value set as you suggested, like below:

```
   nReturned = SQLExecute( hStmt );
   if( nReturned != SQL_SUCCESS ) {
      print_error( SQL_HANDLE_STMT, hStmt, nReturned, "SQLExecute" );
      nReturned = SQLFreeHandle( SQL_HANDLE_STMT, hStmt );
      return 1;   // have an error
   }

   nReturned = SQLMoreResults( hStmt );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( hStmt, nReturned, "SQLFreeHandle(SQL_HANDLE_STMT)" );

   nReturned = SQLMoreResults( hStmt );
   if( nReturned != SQL_SUCCESS )
      print_error_and_exit_stmt( hStmt, nReturned, "SQLFreeHandle(SQL_HANDLE_STMT)" );
```

Unfortunatelly, the table test.drv04_array_dml contains the below rows (see output from chc):

```
chc (ClickHouse CLI portable) v0.1.6
Connecting to database default at localhost:8123 as user test.
Connected to ClickHouse server version 20.10.3.30.

:) select * from test.drv04_array_dml;
┌─ID─┬─c_event_date─┬────dec28_8─┬─dec38_8─┐
│  1 │   1970-01-01 │ 1.00000000 │   1.000 │
│  1 │   1970-01-01 │ 1.00000000 │   1.000 │
│  1 │   1970-01-01 │ 1.00000000 │   1.000 │
└────┴──────────────┴────────────┴─────────┘

3 rows in set. Elapsed: 14.9949ms
```

As you can see, the row set contains tripled first row while expected are three different rows.
Or perhaps, you see something wrong in my code.

Regards,
Marek

**Comment by traceon** [comment]
Could you please try the branch from this PR: #325 ?

**Comment by CMa72** [comment]
Hi,

I've tested the PR #325 and it works fine. The table contains as many rows as it was expected.


But I have a question.
I've tried to use SQLRowCount after each SQLMoreResults, but it always returns 0.
See my code:

```
   SQLLEN nRowCount, nRowCount_all = 0;
   nReturned = SQLExecute( hStmt );
   if( nReturned != SQL_SUCCESS ) {
      print_error( SQL_HANDLE_STMT, hStmt, nReturned, "SQLExecute" );
      nReturned = SQLFreeHandle( SQL_HANDLE_STMT, hStmt );
      return 1;   // have an error
   }

   while( nReturned == SQL_SUCCESS ) {
      nReturned = SQLRowCount( hStmt, &nRowCount );
      if( nReturned != SQL_SUCCESS )
         print_error_and_exit_stmt( hStmt, nReturned, "SQLRowCount" );
      nRowCount_all += nRowCount;

      nReturned = SQLMoreResults( hStmt );
   }
   if( nReturned == SQL_NO_DATA ) {
      // ok - end of iteration
   }
   else {
      print_error_and_exit_stmt( hStmt, nReturned, "SQLMoreResults" );
   }

   // here :: nRowCount_all is always equal to ZERO
```
I expected, that the "nRowCount_all" should be equal to the number of rows INERTEd, but it is not.
Could you please help with this?


Regards,
Marek

**Comment by traceon** [comment]
@CMa72 that's a quirk of the current implementation of the driver and the protocol it uses: it is not able to detect the number of affected rows by `INSERT`/`UPDATE`/'DELETE' and for `SELECT` will report the correct number only after the entire result set was fully fetched. I don't see an easy way to implement it right now. If it is important for you, feel free to open a separate issue, and we will track that there.

**Comment by CMa72** [comment]
Thank you for your support.
Could you provide estimated date of the official release containing this patch?

**Comment by traceon** [comment]
The new version with the fix was released: [v1.1.9.20201226](https://github.com/ClickHouse/clickhouse-odbc/releases/tag/v1.1.9.20201226)
